ic-crypto-getrandom-for-wasm/src/lib.rs

```diff
@@ -19,20 +19,60 @@
 //! See the [getrandom
 //! documentation](https://docs.rs/getrandom/latest/getrandom/macro.register_custom_getrandom.html)
 //! for more details on custom implementations.
+//!
+//! To register this custom implementation, call the function inside of your canister's `init` method
+//! or wherever canister initialization happens, with conditional compilation, like this:
+//!
+//! ```ignore
+//! #[init]
+//! pub fn init(&mut self) {
+//!    #[cfg(target_family = "wasm")]
+//!    ic_crypto_getrandom_for_wasm::register_custom_getrandom();
+//! }
+//! ```
 
 #[cfg(all(
     target_family = "wasm",
     target_vendor = "unknown",
     target_os = "unknown"
 ))]
-/// A getrandom implementation that always fails
-pub fn always_fail(_buf: &mut [u8]) -> Result<(), getrandom::Error> {
-    Err(getrandom::Error::UNSUPPORTED)
-}
+pub use custom_getrandom_impl::register_custom_getrandom;
 
 #[cfg(all(
     target_family = "wasm",
     target_vendor = "unknown",
     target_os = "unknown"
 ))]
-getrandom::register_custom_getrandom!(always_fail);
+mod custom_getrandom_impl {
+    use std::cell::RefCell;
+    use std::time::Duration;
+
+    use candid::Principal;
+    use rand::rngs::StdRng;
+    use rand::{RngCore, SeedableRng};
+
+    thread_local! {
+        static RNG: RefCell<Option<StdRng>> = const { RefCell::new(None) };
+    }
+
+    pub fn register_custom_getrandom() {
```

Comment:
This is open to an entire family of possible bugs and attacks, let's discuss it carefully.

Reply:
For reference: https://forum.dfinity.org/t/seeded-rng-generates-same-random-number-on-every-call/23855

Reply:
> For reference: https://forum.dfinity.org/t/seeded-rng-generates-same-random-number-on-every-call/23855

I may be wrong here, but the concern of the OP in this thread is that he is using the rand setup to generate nonces, which are later consumed. Also, the use of thread-local storage, which ensures that each thread has its own instance of the `RNG`, provides some layer of security for maintaining isolation in a multi-threaded context.

So, it's important that we understand exactly how the `RNG` can be reproduced/predicted. @ufoscout could you please elaborate on this?